### PR TITLE
Instance comment

### PR DESCRIPTION
**Pull request by davidlatwe**
Hi all, it's been a long time :)

## Feature

Commenting on each instance.

312c9b55-3767-4e7c-8809-f32b2de54bce

This feature is requested by my colleague, and I thought this maybe worth to share.

<details>
<summary>Example Code</summary>

```python

import pyblish.api
import pyblish_qml.api


class CreateInstance(pyblish.api.ContextPlugin):
    order = 0
    def process(self, context):
        instance_names = [
            "foo",
            "bar",
        ]
        for name in instance_names:
            context.create_instance(name)


class OpenComment(pyblish.api.ContextPlugin):
    order = 0.1
    def process(self, context):
        context.data["comment"] = ""
        for instance in context:
            instance.data["comment"] = ""


class ValidateContextComment(pyblish.api.ContextPlugin):
    order = 1
    def process(self, context):
        if not context.data.get("comment"):
            self.log.warning("Context has no comment.")


class ValidateInstanceComment(pyblish.api.InstancePlugin):
    order = 1
    def process(self, instance):
        if not instance.data.get("comment"):
            self.log.warning("Instance %r has no comment." % instance.name)


class ComposeComment(pyblish.api.ContextPlugin):
    order = 2
    def process(self, context):
        con_comment = context.data.get("comment", "")
        self.log.debug(con_comment)

        for instance in context:
            int_comment = "%s\n----------\n%s" % (
                con_comment,
                instance.data.get("comment", "")
            )
            instance.data["comment"] = int_comment
            self.log.info(int_comment)


pyblish.api.deregister_all_plugins()
pyblish.api.deregister_all_paths()

pyblish.api.register_plugin(CreateInstance)
pyblish.api.register_plugin(OpenComment)
pyblish.api.register_plugin(ValidateContextComment)
pyblish.api.register_plugin(ValidateInstanceComment)
pyblish.api.register_plugin(ComposeComment)

pyblish_qml.api.show()

```

</details>

## Fix

In commit 8240de4ee87bb470b1d718a80de143f849e28d87, fixed the issue that looks like below. The icon of log message is missing at level like `Info` or `Warning`.

image

### CODE REVIEW

File: pyblish_qml/control.py [hunk 1]
```diff
@@ -665,10 +665,17 @@ def echo(self, data):
...
-        self.host.emit("commented", comment=comment)
+        name = item.name
+        model_item = self.data["models"]["item"].instances[item.id]
+        model_item.hasComment = bool(comment)
+
+        self.host.update(key="comment", value=comment, name=name)
+        if name == "Context":
+            self.host.emit("commented", comment=comment)
+        else:
+            self.host.emit("instanceCommented", comment=comment, name=name)
```
Comment (davidlatwe) [inline pyblish_qml/control.py:678]: Here introduced a new `pyblish` callback signal `instanceCommented`.

**Comment by BigRoy** [comment]
This is looking really nice! :)

File: pyblish_qml/qml/Perspective/Page.qml [hunk 2]
```diff
@@ -8,6 +8,33 @@ Item {
...
+    function setComment(text) {
+        if (overview.state == "") {
+            app.commenting(text, item.name)
+            footer.hasComment = text ? true : false
+        }
+    }
+
+    function getComment() {
+        // GUI may freeze if querying comment while publishing
+        // - note, this triggers commentChanged signal
```
Comment (mottosso) [inline pyblish_qml/qml/Perspective/Page.qml:23]: Maybe we can do a `if old_comment == new_comment: return` to avoid duplicate signals?

**Review by mottosso — COMMENTED** [review]
Very clean code, and nice design. I like it a lot. :)

File: pyblish_qml/qml/Perspective/Page.qml (hunk 2 above)
Comment (davidlatwe) [reply pyblish_qml/qml/Perspective/Page.qml:23]: Ah, that's just like a note-to-self kind of thing, no duplicated signals from here actually. :)

File: pyblish_qml/qml/Perspective/Page.qml [hunk 3]
```diff
@@ -126,6 +153,47 @@ Item {
...
+        readOnly: overview.state != ""
+        text: item.itemType == "instance" ? getComment() : ""
+
+        isUp: root.commenting
+
+        anchors {
+            bottom: footer.top
+            left: parent.left
+            right: parent.right
+            top: (isMaximised && height == parent.height - xfooter.height) ? tabBar.top : undefined
```
Comment (davidlatwe) [inline pyblish_qml/qml/Perspective/Page.qml:171]: Here actually have a typo, `xfooter`, which doesn't exists but yet the code works.
Looks like the value of `isMaximised` is always being `false`, which makes the following conditions as dead code. Fix coming.

**Comment by davidlatwe** [comment]
Really glad you all like this feature !

**Comment by mottosso** [comment]
Happy to pull the trigger once you're happy @davidlatwe, great work.

**Pull request merged**